lib/src/tag_detection.dart

```diff
@@ -693,82 +693,40 @@ class _Sdk {
-
-  _NullSafetyViolationFinder(_PackageGraph packageGraph,
-      _PubspecCache pubspecCache, AnalysisSession analysisSession)
-      : _languageVersionViolationFinder = _PathFinder<String>(packageGraph, (
-          String packageDir,
-        ) {
-          final pubspec = pubspecCache.pubspecOfPackage(packageDir);
-
-          return pubspec.sdkConstraintStatus.hasOptedIntoNullSafety
-              ? null
-              : (path) => Explanation(
-                    'Package is not null safe',
-                    'Because:\n${_PackageGraph.formatPath(path)} '
-                        'that doesn\'t opt in to null safety',
-                    tag: nullSafeTag,
-                  );
-        }),
-        _noOptoutViolationFinder = _PathFinder(
-          packageGraph,
-          (packageName) {
-            for (final file
-                in _dartFilesFromLib(pubspecCache._packageDir(packageName))) {
-              final uri =
-                  Uri.parse('package:$packageName/${path.toUri(file).path}');
-              final unit = _parsedUnitFromUri(analysisSession, uri);
-              if (unit == null) continue;
-              final languageVersionToken = unit.languageVersionToken;
-              if (languageVersionToken == null) continue;
-              if (!isNullSafety(
-                  Version(languageVersionToken.major,
-                      languageVersionToken.minor, 0),
-                  packageName)) {
-                return (path) => Explanation(
-                      'Package is not null safe',
-                      'Because:\n${_PackageGraph.formatPath(path)} where $file is opting out from null-safety.',
-                      tag: nullSafeTag,
-                    );
-              }
-            }
-            return null;
-          },
-        );
-
-  Explanation findNullSafetyViolation(String rootPackageName) {
-    return _languageVersionViolationFinder.findViolation(rootPackageName) ??
-        _noOptoutViolationFinder.findViolation(rootPackageName);
-  }
-
-  static const nullSafeTag = 'is:null-safe';
-}
-
 /// Calculates the tags for the package residing in a given directory.
 class Tagger {
   final String packageName;
   final AnalysisSession _session;
   final _PubspecCache _pubspecCache;
   final bool _isBinaryOnly;
+
+  /// The `lib/<packageName>.dart` library, or `null` if not present.
+  final Uri _primaryLibrary;
+
+  /// All libraries in `lib/` except those in `lib/src/`.
+  final List<Uri> _publicLibraries;
+
+  /// This is:
+  ///  * _primaryLibrary, if one exists,
+  ///  * _publicLibraries, if no _primaryLibrary exists,
+  ///  * all libraries, if no _publicLibraries exists.
+  ///
+  /// TODO(jonasfj): We should be using [_primaryLibrary] (with fallback to
+  ///     [_publicLibraries]) most places instead, and remove this property.
+  ///     If a package has no public libraries, then we should perhaps avoid
+  ///     assigning any tags or assign all tags.
   final List<Uri> _topLibraries;
```

Comment:
This might be orthogonal, but `_topLibraries` seems like a bad name...
For null-safety detection we certainly want to look at "all public libraries".

For platform tag detection we only want to be looking at:
 * the primary library, or
 * the public libraries (if there is no single primary library).
 * MAYBE: all libraries if there is no public libraries -- I'm increasingly leaning towards the idea that we should look at these until someone imports from another library (possibly in another package).

So if a package has no public library, maybe we should just give it all tags or no tags. I'm guessing it's most correct to give it all tags, because it's most likely:
 * (i) A tool that generates code or does other useful things, or,
 * (ii) A package containing private libraries being used by other packages.

In the case of (ii) there is no harm in showing the package.. It's rather useless because it has no public libraries. And of the author doesn't want to pollute search results, the unlisted flag is there for this.